### PR TITLE
VP-6727: Add expired tokens prune hangfire job

### PR DESCRIPTION
**Pull request by akak1977**
There is a possibility for impact to auth token issues performance of oversized OpenIdDict tables.
This PR implements:
1. A hangfire job for pruning expired/invalid tokens.
2. A couple of settings to enable/disable pruning and cron-expression to set up the job start frequency.

image
image

### CODE REVIEW

**Review by tatarincev — COMMENTED** [review]
Try to register the recurring job this way:
https://github.com/VirtoCommerce/vc-platform/blob/master/tests/VirtoCommerce.Platform.Tests/UnitTests/JobSettingsWatcherUnitTests.cs#L32

**Review by OlegoO — APPROVED** [review]

File: src/VirtoCommerce.Platform.Web/Security/BackgroundJobs/PruneExpiredTokensJob.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            _openIddictAuthorizationManager = openIddictAuthorizationManager;
+        }
+
+        [DisableConcurrentExecution(10)]
+        // "DisableConcurrentExecutionAttribute" prevents to start simultaneous job payloads.
+        // Should have short timeout, because this attribute implemented by following manner: newly started job falls into "processing" state immediately.
+        // Then it tries to receive job lock during timeout. If the lock received, the job starts payload.
+        // When the job is awaiting desired timeout for lock release, it stucks in "processing" anyway. (Therefore, you should not to set long timeouts (like 24*60*60), this will cause a lot of stucked jobs and performance degradation.)
+        // Then, if timeout is over and the lock NOT acquired, the job falls into "scheduled" state (this is default fail-retry scenario).
+        // Failed job goes to "Failed" state (by default) after retries exhausted.
```
Comment (ksavosteev) [inline src/VirtoCommerce.Platform.Web/Security/BackgroundJobs/PruneExpiredTokensJob.cs:28]: remove this comment

File: src/VirtoCommerce.Platform.Web/Security/ApplicationBuilderExtensions.cs [hunk 2]
```diff
@@ -38,6 +43,27 @@ public static IApplicationBuilder UseSecurityHandlers(this IApplicationBuilder a
...
+        /// Schedule a periodic job for prune expired/invalid authorization tokens
+        /// </summary>
+        /// <param name="appBuilder"></param>
+        /// <returns></returns>
+        public static IApplicationBuilder UsePruneExpiredTokensJob(this IApplicationBuilder appBuilder)
+        {            
+            var recurringJobManager = appBuilder.ApplicationServices.GetService<IRecurringJobManager>();
+            var settingsManager = appBuilder.ApplicationServices.GetService<ISettingsManager>();
+
+            RecurringJobExtensions.WatchJobSetting(recurringJobManager,
```
Comment (kostyrin) [inline src/VirtoCommerce.Platform.Web/Security/ApplicationBuilderExtensions.cs:56]: please change to recurringJobManager.WatchJobSetting(settingsManager, ...

Comment (akak1977) [reply src/VirtoCommerce.Platform.Web/Security/ApplicationBuilderExtensions.cs]: Agree!

File: src/VirtoCommerce.Platform.Web/Security/ApplicationBuilderExtensions.cs [hunk 3]
```diff
@@ -38,6 +43,27 @@ public static IApplicationBuilder UseSecurityHandlers(this IApplicationBuilder a
...
+            var recurringJobManager = appBuilder.ApplicationServices.GetService<IRecurringJobManager>();
+            var settingsManager = appBuilder.ApplicationServices.GetService<ISettingsManager>();
+
+            recurringJobManager.WatchJobSetting(
+            settingsManager,
+            new SettingCronJobBuilder()
+                .SetEnablerSetting(PlatformConstants.Settings.Security.EnablePruneExpiredTokensJob)
+                .SetCronSetting(PlatformConstants.Settings.Security.CronPruneExpiredTokensJob)
+                .ToJob<PruneExpiredTokensJob>(x => x.Process())
+                .Build());
```
Comment (kostyrin) [inline src/VirtoCommerce.Platform.Web/Security/ApplicationBuilderExtensions.cs:62]: add space

**Review by tatarincev — DISMISSED** [review]

**Pull request merged**